src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/VSInternalCompletionListExtensions.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    internal static class VSInternalCompletionListExtensions
+    {
+        // This needs to match what's listed in VSInternalCompletionItemExtensions.ResultIdKey
+        private const string ResultIdKey = "_resultId";
+
+        public static void SetResultId(
+            this VSInternalCompletionList completionList,
+            int resultId,
+            VSInternalCompletionSetting? completionSetting)
+        {
+            if (completionList is null)
+            {
+                throw new ArgumentNullException(nameof(completionList));
+            }
+
+            var data = new JObject()
+            {
+                [ResultIdKey] = resultId,
+            };
+            if (completionSetting?.CompletionList?.Data == true)
```

Comment:
This is why the `CompletionListOptimizer` doesn't need to care about `Data` anymore